### ISSUE TITLE
Refactor postMessage handling in grain-client.js

### ISSUE THREAD

**Issue by troyjfarrell** [sandstorm-io/sandstorm#3464]
Sandstorm's `postMessage` event listener has more than 230 lines and should be refactored into a set of smaller functions with tests.  If we don't want to move all of it to separate functions, perhaps at least `renderTemplate`, which has 142 lines, can be moved to a separate function.

https://github.com/sandstorm-io/sandstorm/blob/2532bb9608136e454894e276234fe1419b03b405/shell/imports/client/grain-client.js#L1499-L1728

**Comment by troyjfarrell** [comment]
@zenhack Please take a look at https://github.com/troyjfarrell/sandstorm/commit/b37615232695f051a20bf560b44ec16d23f3f11d and let me know if this direction makes sense.  I'm new to TypeScript, so nothing here is going to be great, but the postMessage tests pass.  I don't think that running `make test` gives the TypeScript warnings, so I ran the TypeScript compiler manually.  I'd love to know if there is a better way.

**Comment by zenhack** [comment]
@troyjfarrell, made a couple comments in-line. Didn't read the full code, I assumed you just moved stuff and added annotations, so only really looked at the use of types.

tsc gets run as part of `make lint`, along with eslint. I wish meteor reported type errors as part of the build process (see https://github.com/meteor/meteor/issues/10939)

**Comment by troyjfarrell** [comment]
Yes, I mostly moved stuff and added annotations.  I'll make a second pass, keeping your comments in mind.

**Comment by troyjfarrell** [comment]
I think that this second pass is better.  Please take a look at https://github.com/troyjfarrell/sandstorm/commit/804b2ba8906674888a2e4aacdbc348b333c86d48 at your convenience.  There are still a few TypeScript errors.  I'm not sure how to fix them.  The one that I'd like to fix is the error on `[].concat()' on an array of mixed types.  Maybe it needs to accept an `any` array?  I'm not worried about the others, but I also think that it would be better not to commit code that adds TypeScript errors.

**Comment by zenhack** [comment]
@troyjfarrell, Re: the concat error I assume you're talking about:

```javascript
Meteor.call.apply(Meteor, ["newApiToken"].concat(params, callback));
```

I don't really understand why that's not just:

```javascript
Meteor.call("newApiToken", params, callback)
```

...which might typecheck on its own?

@kentonv, any insight re: what's going on with that line of code?

---

I left some comments in-line. if/when you do another pass, could you open a proper pr (marked as a draft if it's not ready for merging)? It's a bit hard to keep multiple review passes organized this way.

We should definitely merging type errors; I'd like to be able to keep CI failing if they crop up. If there's something we can't make fit due to type-unfriendly constructs we can short-term use `any` or the like, but we shouldn't leave the types not checking.

Most of the other type errors seem to be due to not having proper type declarations for some dependency; this is why I've tried to start converting stuff at the leaves of the dependency graph. So maybe we're stuck using any or a cast for some of those, but I'd add comments with a TODO to get rid of that once types for the dependency are available.
